cacheops/redis.py

```diff
@@ -25,115 +42,74 @@ def handle_connection_failure(call):
-                raise
-            connection = self.connection_pool.get_connection(args[0], **options)
-            connection.disconnect()
-            set_read_clients()
-            warnings.warn("Primary probably failed over, reconnecting")
-            return super(SafeRedis, self).execute_command(*args, **options)
-
-
-class LazyRedis(object):
-    def _setup(self):
-        Redis = SafeRedis if settings.CACHEOPS_DEGRADE_ON_FAILURE else redis.StrictRedis
-
-        # Allow client connection settings to be specified by a URL.
-        if isinstance(settings.CACHEOPS_REDIS, six.string_types):
-            client = Redis.from_url(settings.CACHEOPS_REDIS)
-        else:
-            client = Redis(**settings.CACHEOPS_REDIS)
-
-        object.__setattr__(self, '__class__', client.__class__)
-        object.__setattr__(self, '__dict__', client.__dict__)
-
-    def __getattr__(self, name):
-        self._setup()
-        return getattr(self, name)
-
-    def __setattr__(self, name, value):
-        self._setup()
-        return setattr(self, name, value)
-
-CacheopsRedis = SafeRedis if settings.CACHEOPS_DEGRADE_ON_FAILURE else client_class
-try:
-    set_read_clients()
-except AttributeError as err:
-    redis_client = LazyRedis()
-else:
-    class ReplicaProxyRedis(CacheopsRedis):
-        """Proxy `get` calls to redis replica."""
         def get(self, *args, **kwargs):
+            """Proxy `get` calls to redis replica."""
+            if not self.read_clients:
+                self.set_read_clients()
             try:
-                client = random.choice(read_clients)
+                client = random.choice(self.read_clients)
                 return client.get(*args, **kwargs)
             except redis.ConnectionError:
                 return super(ReplicaProxyRedis, self).get(*args, **kwargs)
 
-    if isinstance(settings.CACHEOPS_REDIS, six.string_types):
-        redis_client = ReplicaProxyRedis.from_url(settings.CACHEOPS_REDIS)
-    else:
-        redis_client = ReplicaProxyRedis(**settings.CACHEOPS_REDIS)
+        def execute_command(self, *args, **options):
+            """Handle failover of AWS elasticache."""
+            try:
+                return super(SafeRedis, self).execute_command(*args, **options)
+            except redis.ResponseError as e:
+                if "READONLY" not in e.message:
+                    raise
+                connection = self.connection_pool.get_connection(args[0], **options)
+                connection.disconnect()
+                self.read_clients = []
+                warnings.warn("Primary probably failed over, reconnecting")
+                return super(SafeRedis, self).execute_command(*args, **options)
 
-### Lua script loader
+    client_class = ReplicaProxyRedis
 
-import re
-import os.path
+
+if isinstance(settings.CACHEOPS_REDIS, six.string_types):
+    redis_client = client_class.from_url(settings.CACHEOPS_REDIS)
+else:
+    redis_client = client_class(**settings.CACHEOPS_REDIS)
```

Comment:
In theory we should still be handling Redis URLs as well as dicts – we can clean that up after we update our fork onto cacheops 4.